File: Zeze/Net/ServiceConf.cs
using System;
using System.Collections.Generic;
using System.Xml;
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using Zeze.Services;
using System.Threading.Tasks;
using Zeze.Util;

namespace Zeze.Net
{
    public sealed class ServiceConf
    {
        public Service Service { get; private set; }
        public readonly string Name;
        public SocketOptions SocketOptions = new SocketOptions();
        public HandshakeOptions HandshakeOptions = new HandshakeOptions();

        private readonly ConcurrentDictionary<string, Acceptor> Acceptors
            = new ConcurrentDictionary<string, Acceptor>();

        private readonly ConcurrentDictionary<string, Connector> Connectors
            = new ConcurrentDictionary<string, Connector>();

        internal void SetService(Service service)
        {
            lock (this)
            {
                if (Service != null)
                    throw new Exception($"ServiceConf of '{Name}' Service != null");
                Service = service;
                ForEachAcceptor(a => a.SetService(service));
                ForEachConnector(c => c.SetService(service));
            }
        }

        public void AddConnector(Connector connector)
        {
            connector.SetService(Service);
            if (!Connectors.TryAdd(connector.Name, connector))
                throw new Exception($"Duplicate Connector={connector.Name}");
        }

        public Connector FindConnector(string name)
        {
            return Connectors.TryGetValue(name, out var exist) ? exist : null;
        }

        public Connector FindConnector(string host, int port)
        {
            return FindConnector($"{host}:{port}");
        }

        /// <summary>
        /// 查找，不存在则创建。
        /// </summary>
        /// <param name="host"></param>
        /// <param name="port"></param>
        /// <param name="autoReconnect"></param>
        /// <param name="getOrAdd"></param>
        /// <returns>true if addNew</returns>
        public bool TryGetOrAddConnector(string host, int port, bool autoReconnect, out Connector getOrAdd)
        {
            bool addNew = false;
            getOrAdd = Connectors.GetOrAdd($"{host}:{port}", _ =>
            {
                var add = new Connector(host, port, autoReconnect);
                add.SetService(Service);
                addNew = true;
                return add;
            });
            return addNew;
        }

        public void RemoveConnector(Connector c)
        {
            Connectors.TryRemove(new KeyValuePair<string, Connector>(c.Name, c));
        }

        public void ForEachConnector(Action<Connector> action)
        {
            foreach (var c in Connectors.Values)
                action(c);
        }

        public async Task ForEachConnectorAsync(Func<Connector, Task> action)
        {
            foreach (var c in Connectors.Values)
                await action(c);
        }

        public int ConnectorCount()
        {
            return Connectors.Count;
        }

        public bool ForEachConnector(Func<Connector, bool> func)
        {
            foreach (var c in Connectors.Values)
            {
                if (!func(c))
                    return false;
            }
            return true;
        }

        public void AddAcceptor(Acceptor a)
        {
            if (!Acceptors.TryAdd(a.Name, a))
                throw new Exception($"Duplicate Acceptor={a.Name}");
            a.SetService(Service);
        }

        public void RemoveAcceptor(Acceptor a)
        {
            Acceptors.TryRemove(new KeyValuePair<string, Acceptor>(a.Name, a));
        }

        public void ForEachAcceptor(Action<Acceptor> action)
        {
            foreach (var a in Acceptors.Values)
                action(a);
        }

        public bool ForEachAcceptor(Func<Acceptor, bool> func)
        {
            foreach (var a in Acceptors.Values)
            {
                if (!func(a))
                    return false;
            }
            return true;
        }

        public int AcceptorCount()
        {
            return Acceptors.Count;
        }

        public ServiceConf()
        {
            Name = "";
        }

        public ServiceConf(Config conf, XmlElement self)
        {
            Name = self.GetAttribute("Name");

            // SocketOptions
            var attr = self.GetAttribute("NoDelay");
            if (attr.Length > 0) SocketOptions.NoDelay = bool.Parse(attr);
            attr = self.GetAttribute("SendBuffer");
            if (attr.Length > 0) SocketOptions.SendBuffer = int.Parse(attr);
            attr = self.GetAttribute("ReceiveBuffer");
            if (attr.Length > 0) SocketOptions.ReceiveBuffer = int.Parse(attr);
            attr = self.GetAttribute("InputBufferSize");
            if (attr.Length > 0) SocketOptions.InputBufferSize = int.Parse(attr);
            attr = self.GetAttribute("InputBufferMaxProtocolSize");
            if (attr.Length > 0) SocketOptions.InputBufferMaxProtocolSize = int.Parse(attr);
            attr = self.GetAttribute("OutputBufferMaxSize");
            if (attr.Length > 0) SocketOptions.OutputBufferMaxSize = int.Parse(attr);
            attr = self.GetAttribute("Backlog");
            if (attr.Length > 0) SocketOptions.Backlog = int.Parse(attr);
            attr = self.GetAttribute("SocketLogLevel");
            if (attr.Length > 0)
                SocketOptions.SocketLogLevel = (LogLevel)Enum.Parse(typeof(LogLevel), attr);
            attr = self.GetAttribute("TimeThrottleSeconds");
            if (attr.Length > 0) SocketOptions.TimeThrottleSeconds = int.Parse(attr);
            attr = self.GetAttribute("TimeThrottleLimit");
            if (attr.Length > 0) SocketOptions.TimeThrottleLimit = int.Parse(attr);
            attr = self.GetAttribute("TimeThrottleBandwidth");
            if (attr.Length > 0) SocketOptions.TimeThrottleBandwidth = int.Parse(attr);
            attr = self.GetAttribute("OverBandwidth");
            if (attr.Length > 0) SocketOptions.OverBandwidth = long.Parse(attr);
            attr = self.GetAttribute("OverBandwidthFusingRate");
            if (attr.Length > 0) SocketOptions.OverBandwidthFusingRate = double.Parse(attr);
            attr = self.GetAttribute("OverBandwidthNormalRate");
            if (attr.Length > 0) SocketOptions.OverBandwidthNormalRate = double.Parse(attr);

            // HandshakeOptions
            attr = self.GetAttribute("DhGroups");
            if (attr.Length > 0)
            {
                HandshakeOptions.DhGroups = new HashSet<int>();
                foreach (string dg in attr.Split(','))
                {
                    string dgTmp = dg.Trim();
                    if (dgTmp.Length == 0)
                        continue;
                    HandshakeOptions.AddDhGroup(int.Parse(dgTmp));
                }
            }
            attr = self.GetAttribute("SecureIp");
            if (attr.Length > 0) HandshakeOptions.SecureIp = IPAddress.Parse(attr).GetAddressBytes();
            attr = self.GetAttribute("RsaPubKey");
            if (attr.Length > 0) HandshakeOptions.LoadRsaPubKey(Str.toBytes(attr));
            attr = self.GetAttribute("RsaPriKeyFile");
            if (attr.Length > 0) HandshakeOptions.LoadRsaPriKey(File.ReadAllBytes(attr));
            attr = self.GetAttribute("CompressS2c");
            if (attr.Length > 0) HandshakeOptions.CompressS2c = int.Parse(attr);
            attr = self.GetAttribute("CompressC2s");
            if (attr.Length > 0) HandshakeOptions.CompressC2s = int.Parse(attr);
            attr = self.GetAttribute("EncryptType");
            if (attr.Length > 0) HandshakeOptions.EncryptType = int.Parse(attr);

            attr = self.GetAttribute("KeepCheckPeriod");
            if (attr.Length > 0) HandshakeOptions.KeepCheckPeriod = int.Parse(attr);
            attr = self.GetAttribute("KeepSendTimeout");
            if (attr.Length > 0) HandshakeOptions.KeepSendTimeout = int.Parse(attr);
            attr = self.GetAttribute("KeepRecvTimeout");
            if (attr.Length > 0) HandshakeOptions.KeepRecvTimeout = int.Parse(attr);

            if (string.IsNullOrEmpty(Name))
                conf.DefaultServiceConf = this;
            else if (!conf.ServiceConfMap.TryAdd(Name, this))
                throw new Exception($"Duplicate ServiceConf '{Name}'");

            // connection creator options
            foreach (XmlNode node in self.ChildNodes)
            {
                if (node.NodeType != XmlNodeType.Element)
                    continue;

                var e = (XmlElement)node;
                switch (e.Name)
                {
                    case "Acceptor":
                        AddAcceptor(new Acceptor(e));
                        break;
                    case "Connector":
                        AddConnector(Connector.Create(e));
                        break;
                    default: throw new Exception("unknown node name: " + e.Name);
                }
            }
        }

        public void Start()
        {
            ForEachAcceptor(a => a.Start());
            ForEachConnector(c => c.Start());
        }

        public void Stop()
        {
            ForEachAcceptor(a => a.Stop());
            ForEachConnector(c => c.Stop());
        }

        public void StopListen()
        {
            ForEachAcceptor(a => a.Stop());
        }
    }
}

File: ZezeJava/ZezeJava/src/main/java/Zeze/Net/ServiceConf.java
package Zeze.Net;

import java.net.InetAddress;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Predicate;
import Zeze.Config;
import Zeze.Services.HandshakeOptions;
import Zeze.Util.Action1;
import Zeze.Util.Func3;
import Zeze.Util.IntHashSet;
import Zeze.Util.OutObject;
import Zeze.Util.Str;
import Zeze.Util.Task;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

public final class ServiceConf extends ReentrantLock {
	private Service service;
	private final @NotNull String name;
	private @NotNull SocketOptions socketOptions = new SocketOptions();
	private @NotNull HandshakeOptions handshakeOptions = new HandshakeOptions();
	private final ConcurrentHashMap<String, Acceptor> acceptors = new ConcurrentHashMap<>();
	private final ConcurrentHashMap<String, Connector> connectors = new ConcurrentHashMap<>();
	private int maxConnections = 1024; // 适合绝大多数网络服务，对于连接机，比如Linkd，Gated等需要自己加大。

	public Service getService() {
		return service;
	}

	public @NotNull String getName() {
		return name;
	}

	public int getMaxConnections() {
		return maxConnections;
	}

	public @NotNull SocketOptions getSocketOptions() {
		return socketOptions;
	}

	public void setSocketOptions(@NotNull SocketOptions value) {
		//noinspection ConstantValue
		if (value != null)
			socketOptions = value;
	}

	public @NotNull HandshakeOptions getHandshakeOptions() {
		return handshakeOptions;
	}

	public void setHandshakeOptions(@NotNull HandshakeOptions value) {
		//noinspection ConstantValue
		if (value != null)
			handshakeOptions = value;
	}

	public void setService(Service service) {
		lock();
		try {
			if (this.service != null) {
				throw new IllegalStateException(String.format("ServiceConf of '%s' Service != null", getName()));
			}
			this.service = service;
			forEachAcceptor(a -> a.SetService(service));
			forEachConnector(c -> c.SetService(service));
		} finally {
			unlock();
		}
	}

	public void addConnector(@NotNull Connector connector) {
		if (null != connectors.putIfAbsent(connector.getName(), connector)) {
			throw new IllegalStateException("Duplicate Connector=" + connector.getName());
		}
		connector.SetService(service);
	}

	public @Nullable Connector findConnector(@NotNull String name) {
		return connectors.get(name);
	}

	public @Nullable Connector findConnector(@NotNull String host, int port) {
		return findConnector(host + "_" + port);
	}

	/**
	 * 查找，不存在则创建。
	 *
	 * @param host          peer address or name
	 * @param port          peer port
	 * @param autoReconnect auto reconnect when socket close.
	 * @param getOrAdd      out. connector returned.
	 * @return true if addNew
	 */
	public boolean tryGetOrAddConnector(@NotNull String host, int port, boolean autoReconnect,
										@Nullable OutObject<Connector> getOrAdd) {
		return tryGetOrAddConnector(host, port, autoReconnect, getOrAdd, Connector::new);
	}

	public boolean tryGetOrAddConnector(@NotNull String host, int port, boolean autoReconnect,
										@Nullable OutObject<Connector> getOrAdd,
										Func3<String, Integer, Boolean, Connector> cf) {
		var name = host + "_" + port;
		final var addNew = new OutObject<Connector>();
		var c = connectors.computeIfAbsent(name, __ -> {
			Connector add;
			try {
				add = cf.call(host, port, autoReconnect);
			} catch (Exception e) {
				Task.forceThrow(e);
				return null; // never run here
			}
			add.SetService(service);
			addNew.value = add;
			return add;
		});
		if (getOrAdd != null)
			getOrAdd.value = c;
		return addNew.value != null;
	}

	public void removeConnector(@NotNull Connector c) {
		connectors.remove(c.getName(), c);
	}

	public void ForEachConnector(@NotNull Action1<Connector> action) throws Exception {
		for (var c : connectors.values()) {
			action.run(c);
		}
	}

	public void forEachConnector(@NotNull Consumer<Connector> action) {
		for (var a : connectors.values())
			action.accept(a);
	}

	public int connectorCount() {
		return connectors.size();
	}

	public boolean forEachConnector2(@NotNull Predicate<Connector> func) {
		for (var c : connectors.values()) {
			if (!func.test(c))
				return false;
		}
		return true;
	}

	public void addAcceptor(@NotNull Acceptor a) {
		if (null != acceptors.putIfAbsent(a.getName(), a)) {
			throw new IllegalStateException("Duplicate Acceptor=" + a.getName());
		}
		a.SetService(service);
	}

	public void removeAcceptor(@NotNull Acceptor a) {
		acceptors.remove(a.getName(), a);
	}

	public void ForEachAcceptor(@NotNull Action1<Acceptor> action) throws Exception {
		for (var a : acceptors.values())
			action.run(a);
	}

	public void forEachAcceptor(@NotNull Consumer<Acceptor> action) {
		for (var a : acceptors.values())
			action.accept(a);
	}

	public boolean forEachAcceptor2(@NotNull Function<Acceptor, Boolean> func) {
		for (var a : acceptors.values()) {
			if (!func.apply(a))
				return false;
		}
		return true;
	}

	public int acceptorCount() {
		return acceptors.size();
	}

	public ServiceConf() {
		name = "";
	}

	public ServiceConf(@NotNull Config conf, @NotNull Element self) {
		name = self.getAttribute("Name").trim();

		String attr;

		// SocketOptions
		attr = self.getAttribute("NoDelay");
		if (!attr.isBlank())
			getSocketOptions().setNoDelay(Boolean.parseBoolean(attr));

		attr = self.getAttribute("SendBuffer");
		if (!attr.isBlank())
			getSocketOptions().setSendBuffer(Str.parseIntSize(attr));

		attr = self.getAttribute("ReceiveBuffer");
		if (!attr.isBlank())
			getSocketOptions().setReceiveBuffer(Str.parseIntSize(attr));

		attr = self.getAttribute("InputBufferMaxProtocolSize");
		if (!attr.isBlank())
			getSocketOptions().setInputBufferMaxProtocolSize(Str.parseIntSize(attr));

		attr = self.getAttribute("OutputBufferMaxSize");
		if (!attr.isBlank())
			getSocketOptions().setOutputBufferMaxSize(Str.parseLongSize(attr));

		attr = self.getAttribute("Backlog");
		if (!attr.isBlank())
			getSocketOptions().setBacklog(Integer.parseInt(attr));

		getSocketOptions().setTimeThrottle(self.getAttribute("TimeThrottle"));

		attr = self.getAttribute("TimeThrottleSeconds");
		if (!attr.isBlank())
			getSocketOptions().setTimeThrottleSeconds(Integer.parseInt(attr));

		attr = self.getAttribute("TimeThrottleLimit");
		if (!attr.isBlank())
			getSocketOptions().setTimeThrottleLimit(Integer.parseInt(attr));

		attr = self.getAttribute("TimeThrottleBandwidth");
		if (!attr.isBlank())
			getSocketOptions().setTimeThrottleBandwidth(Str.parseIntSize(attr));

		attr = self.getAttribute("OverBandwidth");
		if (!attr.isBlank())
			getSocketOptions().setOverBandwidth(Str.parseLongSize(attr));

		attr = self.getAttribute("OverBandwidthFusingRate");
		if (!attr.isBlank())
			getSocketOptions().setOverBandwidthFusingRate(Double.parseDouble(attr));
		attr = self.getAttribute("OverBandwidthNormalRate");
		if (!attr.isBlank())
			getSocketOptions().setOverBandwidthNormalRate(Double.parseDouble(attr));

		attr = self.getAttribute("CloseWhenMissHandle");
		if (!attr.isBlank())
			getSocketOptions().setCloseWhenMissHandle(Boolean.parseBoolean(attr));

		// HandshakeOptions
		attr = self.getAttribute("DhGroups");
		if (!attr.isBlank()) {
			getHandshakeOptions().setDhGroups(new IntHashSet());
			for (String dg : attr.split(",", -1)) {
				String dgTmp = dg.strip();
				if (!dgTmp.isEmpty())
					getHandshakeOptions().addDhGroup(Integer.parseInt(dgTmp));
			}
		}
		attr = self.getAttribute("SecureIp");
		if (!attr.isBlank()) {
			try {
				getHandshakeOptions().setSecureIp(InetAddress.getByName(attr).getAddress());
			} catch (Exception ex) {
				throw new IllegalStateException(ex);
			}
		}
		attr = self.getAttribute("CompressS2c");
		if (!attr.isBlank())
			getHandshakeOptions().setCompressS2c(Integer.parseInt(attr));

		attr = self.getAttribute("CompressC2s");
		if (!attr.isBlank())
			getHandshakeOptions().setCompressC2s(Integer.parseInt(attr));

		attr = self.getAttribute("EncryptType");
		if (!attr.isBlank())
			getHandshakeOptions().setEncryptType(Integer.parseInt(attr));

		attr = self.getAttribute("maxConnections");
		if (!attr.isBlank())
			maxConnections = Integer.parseInt(attr);

		attr = self.getAttribute("KeepCheckPeriod");
		if (!attr.isBlank())
			getHandshakeOptions().setKeepCheckPeriod(Integer.parseInt(attr));

		attr = self.getAttribute("KeepRecvTimeout");
		if (!attr.isBlank())
			getHandshakeOptions().setKeepRecvTimeout(Integer.parseInt(attr));

		attr = self.getAttribute("KeepSendTimeout");
		if (!attr.isBlank())
			getHandshakeOptions().setKeepSendTimeout(Integer.parseInt(attr));

		{
			String name = getName();
			if (name.isBlank()) {
				conf.setDefaultServiceConf(this);
			} else if (null != conf.getServiceConfMap().putIfAbsent(name, this)) {
				throw new IllegalStateException("Duplicate ServiceConf " + getName());
			}
		}

		// connection creator options
		NodeList childNodes = self.getChildNodes();
		for (int i = 0; i < childNodes.getLength(); ++i) {
			Node node = childNodes.item(i);
			if (Node.ELEMENT_NODE != node.getNodeType()) {
				continue;
			}

			Element e = (Element)node;
			switch (e.getNodeName()) {
			case "Acceptor":
				addAcceptor(new Acceptor(e));
				break;
			case "Connector":
				addConnector(Connector.Create(e));
				break;
			default:
				throw new IllegalStateException("unknown node name: " + e.getNodeName());
			}
		}
	}

	public void start() {
		forEachAcceptor(Acceptor::Start);
		forEachConnector(Connector::start);
	}

	public void stop() {
		forEachAcceptor(Acceptor::Stop);
		forEachConnector(Connector::stop);
	}

	public void stopListen() {
		forEachAcceptor(Acceptor::Stop);
	}
}
